### PR TITLE
Adding encoding flag and support for base64 data

### LINKED ISSUE #1204: Add dataBase64 flag for Ping Source
### Feature request

For parity to be able to create a ping source using the kn cli

### Use case
Use the `dataBase64` for binary data as shown in the [docs](https://knative.dev/docs/eventing/samples/ping-source/)

`kn source ping create test-ping-source --schedule "*/2 * * * *" --dataBase64: "ZGF0YQ==" --sink ksvc:event-display`

### PR DESCRIPTION
**Pull request by vyasgun**
## Description
Adding encoding option to specify base64 or text values. Adding support for base64 data through CLI

## Changes

* Adding encoding option to specify the type of encoding of data.
* Adding code to identify automatically (without any encoding option specified) if given data is base64 encoded or not and setting the relevant field in PingSourceSpec.
* Adding a test case to cover these scenarios.

## Reference

Fixes #1204

### CODE REVIEW

**Comment by vyasgun** [comment]
Looking into the failed integration tests.

File: pkg/kn/commands/source/ping/flags.go [hunk 1]
```diff
@@ -42,6 +43,8 @@ func (c *pingUpdateFlags) addFlags(cmd *cobra.Command) {
 
 	cmd.Flags().StringVarP(&c.data, "data", "d", "", "Json data to send")
 
+	cmd.Flags().StringVarP(&c.encoding, "encoding", "e", "", "Preferred encoding format. Valid values: text/base64")
```
Comment (dsimansk) [inline pkg/kn/commands/source/ping/flags.go:46]: I'd suggest to change the description a bit. `Data encoding format. One of: text|base64.`

File: pkg/sources/v1beta2/ping_client.go [hunk 2]
```diff
@@ -175,6 +175,11 @@ func (b *PingSourceBuilder) Data(data string) *PingSourceBuilder {
 	return b
 }
 
+func (b *PingSourceBuilder) DataBase64(data string) *PingSourceBuilder {
```
Comment (dsimansk) [inline pkg/sources/v1beta2/ping_client.go:178]: The new build function should be added to some test cases.

File: pkg/kn/commands/source/ping/flags.go (hunk 1 above)
Comment (dsimansk) [reply pkg/kn/commands/source/ping/flags.go:46]: In addition it should be also mentioned that `--data` is smarter now and detects base64 encoding.

File: pkg/kn/commands/source/ping/create.go [hunk 3]
```diff
@@ -70,9 +76,29 @@ func NewPingCreateCommand(p *commands.KnParams) *cobra.Command {
 			}
 			ceOverridesToRemove := util.ParseMinusSuffix(ceOverridesMap)
 
+			var data, dataBase64 string
+
+			switch updateFlags.encoding {
```
Comment (dsimansk) [inline pkg/kn/commands/source/ping/create.go:81]: IMO the same handling of `--encoding` flag should be added to `source ping update` as well. Probably extracting the code to `flags.go` to be shared.

Comment (dsimansk) [reply pkg/kn/commands/source/ping/create.go:81]: The `update` op should be able to handle switching between `data` and `dataBase64` field to clear the old one and change to the user's specified.

File: pkg/kn/commands/source/ping/create.go [hunk 4]
```diff
@@ -70,9 +76,29 @@ func NewPingCreateCommand(p *commands.KnParams) *cobra.Command {
...
+					dataBase64 = updateFlags.data
+				} else {
+					data = updateFlags.data
+				}
+			case base64Encoding:
+				dataBase64 = updateFlags.data
+			case textEncoding:
+				data = updateFlags.data
+			default:
+				err := fmt.Errorf("invalid value: %v", updateFlags.encoding)
```
Comment (dsimansk) [inline pkg/kn/commands/source/ping/create.go:93]: Lets instruct user about accepted values `text|base64` in the error message.

**Review by dsimansk — APPROVED** [review]
Thanks @vyasgun! There're some nits to polish, otherwise it looks really good.

**Comment by dsimansk** [comment]
@vyasgun ups I missclicked on approved review instread of require changes. :)
Can you please follow up on my review comments in the next PR.

**Comment by vyasgun** [comment]
@dsimansk Thanks for the review! I will open another PR for update command with these comments addressed.

**Comment by vyasgun** [comment]
@dsimansk I have addressed the comments in #1392

**Pull request merged**